src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java

```diff
@@ -283,328 +340,433 @@ public boolean isPerSSTableBuildComplete()
+        @Override
+        public boolean validateComponents(boolean validateChecksum)
+        {
+            if (isEmpty())
+                return true;
 
-        registerPerSSTableComponent(component);
+            boolean isValid = true;
+            for (IndexComponentType expected : expectedComponentsForVersion())
+            {
+                var component = components.get(expected);
+                if (component == null)
+                {
+                    logger.warn(logMessage("Missing index component {} from SSTable {}"), expected, descriptor);
+                    isValid = false;
+                }
+                else if (!version().onDiskFormat().validateIndexComponent(component, validateChecksum))
+                {
+                    logger.warn(logMessage("Invalid/corrupted component {} for SSTable {}"), expected, descriptor);
+                    if (CassandraRelevantProperties.DELETE_CORRUPT_SAI_COMPONENTS.getBoolean())
+                    {
+                        // We delete the corrupted file. Yes, this may break ongoing reads to that component, but
+                        // if something is wrong with the file, we're rather fail loudly from that point on than
+                        // risking reading and returning corrupted data.
+                        deleteComponentFile(component.file());
+                        // Note that invalidation will also delete the completion marker
+                    }
+                    else
+                    {
+                        logger.debug("Leaving believed-corrupt component {} of SSTable {} in place because {} is false", expected, descriptor, CassandraRelevantProperties.DELETE_CORRUPT_SAI_COMPONENTS.getKey());
+                    }
+
+                    isValid = false;
+                }
+            }
+            if (!isValid)
+                invalidate();
+            return isValid;
+        }
 
-        return writer;
-    }
+        @Override
+        public void invalidate()
+        {
+            // We delete the completion marker, to make it clear the group of components shouldn't be used anymore,
+            // in particular for the following "populate" call. Note it's comparatively safe to do so in that the
+            // marker is never accessed during reads, so we cannot break ongoing operations here.
+            var marker = components.remove(completionMarkerComponent());
+            if (marker != null)
+                deleteComponentFile(marker.file());
+
+            // Keeping legacy behavior if immutable components is disabled.
+            if (!version.useImmutableComponentFiles() && CassandraRelevantProperties.DELETE_CORRUPT_SAI_COMPONENTS.getBoolean())
+                forceDeleteAllComponents();
+
+            groups.remove(context);
+            populateComponents(IndexDescriptor.this, context);
```

Comment:
Completion marker file is removed above.  `populateComponents` will still find completion marker in TOC and populate `IndexComponentsImpl` with latest generation in the following code block:

```
                IndexComponentsImpl components = descriptor.new IndexComponentsImpl(context, version, generation);
                foundTypes.forEach(components::addOrGet);
                components.isComplete = true;
                descriptor.groups.put(context, components);
```

I was expecting it to create `generation=-1`. Is it expected?

Reply:
You're right, we need to unregister the components if they are invalid. Added it (forced to pass a few new arguments but not too bad).